Setup error screens

beausmith opened the pull request:
Fixes #1041 

This PR needs:
- [x] backend data to enable screens.
- [x] tests based upon mocked data.

# Error Screens

These screens will display for all users if the state is such.

## Low battery and no charger
![image](https://user-images.githubusercontent.com/28444/72029195-01541100-323a-11ea-97f0-d50f4caeec77.png)

## Card Reader not detected
![image](https://user-images.githubusercontent.com/28444/72029217-17fa6800-323a-11ea-8afb-2a0141932f07.png)

## No Printer Detected
![image](https://user-images.githubusercontent.com/28444/72045743-b521c480-326b-11ea-8864-e38db6d64802.png)

## Printer Error Detected
![image](https://user-images.githubusercontent.com/28444/72045817-df738200-326b-11ea-8a3a-2670a86e8a7d.png)

# Warnings on Insert Card Screen

These warnings are informational but non-blocking. They are visible to both Poll Workers and Voters.

Note: It is possible for both of these warnings to appear at the same time.

## No Power Warning
![image](https://user-images.githubusercontent.com/28444/72111144-ca8c0280-32ee-11ea-8aa0-882c637cfe00.png)

## No Accessible Controller Warning
![image](https://user-images.githubusercontent.com/28444/72111187-e2fc1d00-32ee-11ea-91d7-bccc016cc754.png)

eventualbuddha commented on src/AppRoot.tsx:
```diff
@@ -793,11 +841,30 @@ class AppRoot extends React.Component<Props, State> {
       userSettings,
       votes,
     } = this.state
+    if (hasLowBattery && !hasChargerAttached) {
```
All these will stop a voter who is in the middle of voting, right? Is that the behavior we want?

beausmith replied:
Yes. For loss of any functionality which will not allow a voter to complete their vote, we display one of these error screens. 

In the case of this specific error, if the charger is not plugged in and `!hasLowBattery`, the voter will see a non-blocking notification warning (TBD) on the Insert Card screen which is informational is design and has copy like: "This machine is on battery power. Please ask a poll worker to plug-in the power cord for this machine."

eventualbuddha commented on src/AppRoot.tsx:
```diff
@@ -456,6 +471,37 @@ class AppRoot extends React.Component<Props, State> {
+  }
+
+  public stopStatusPolling = () => {
+    this.statusPoller && this.statusPoller.stop()
```
```suggestion
    this.statusPoller?.stop()
```

beausmith replied:
Oh yeah! This was copy/pasta/edit, but should have caught it.

beausmith commented:
@eventualbuddha - I have added a Hardware API. This seemed simpler than separate APIs for each. 

Open to all suggestions for any refactors, etc.

What else do you need from me to implement a "module-hardware" or whatever you think is best?

beausmith commented on src/utils/Hardware.ts:
```diff
@@ -0,0 +1,246 @@
+    AcessibleControllerAPI
+  > {
+    console.log("Why is 'this' undefined?", this)
+    return this.accessibleControllerStatus
```
@eventualbuddha - The last commit on this branch is a WIP with debugging code. I'm stumped as to why `this` is undefined. `MemoryHardware` is modeled after the `MemoryCard`, but something must be different. Do you see where the error in this code is?

eventualbuddha replied:
Commented at the source of the problem. It’s an unbound method problem. I’d fix it by doing `const { hardware } = this.props` and calling i.e. `await hardware.read…`.

eventualbuddha commented on src/AppRoot.tsx:
```diff
@@ -456,6 +472,55 @@ class AppRoot extends React.Component<Props, State> {
+              readCardReaderStatus,
+              readPrinterStatus,
+            } = this.props.hardware
+            const accesssibleController = await readAccesssibleControllerStatus()
```
This is the call that’s making `this` be `undefined` inside `readAccesssibleControllerStatus`. When you call a function without a context in strict mode you get `this === undefined` (in non-strict mode you get `this === window`).

For example:
```ts
const obj = { m() { return this } }
const { m } = obj
const mBound = ::obj.m // https://github.com/tc39/proposal-bind-operator

// These all pass.
assert(obj.m() === obj)
assert(m() === undefined)
assert(m.call(obj) === obj)
assert(mBound() === obj)
```

Note that the bind operator proposal is still stage 0 and [won’t be implemented in TS until stage 3 at the earliest](https://github.com/Microsoft/TypeScript/issues/3508).

eventualbuddha commented on src/utils/Hardware.ts:
```diff
@@ -0,0 +1,246 @@
+  /**
+   * Sets Battery level.  Number between 0–1.
+   */
+  setBatteryLevel(level: number): Promise<void>
```
These `set*` methods don’t represent operations we can actually perform in non-test scenarios, i.e. the battery level is what it is. I’d say these should not be present on the `Hardware` interface but should be defined on `MemoryHardware`. Unless I’m missing something?

beausmith commented:
@eventualbuddha - I've rebased, squashed, and pushed commits. Everything is working, however there are a couple regressions I've failed to figure out. I've got to hang my hat for the night. Back to this tomorrow.
